pythonforandroid/graph.py

```diff
@@ -99,17 +132,129 @@ def find_order(graph):
+        recipe order or choosing an actual value in any of the multiple
+        choice tuples/dependencies, and just do a very basic obvious
+        conflict check.
+    """
+    deps_were_added_by = dict()
+    deps = set()
+
+    # Add dependencies for all recipes:
+    to_be_added = [(name_tuple, None) for name_tuple in name_tuples]
+    while len(to_be_added) > 0:
+        current_to_be_added = list(to_be_added)
+        to_be_added = []
+        for (added_tuple, adding_recipe) in current_to_be_added:
+            assert(type(added_tuple) == tuple)
+            if len(added_tuple) > 1:
+                # No obvious commitment in what to add, ignore it.
+                # (Remember this function only catches obvious issues)
+                continue
+
+            name = added_tuple[0]
+            recipe_conflicts = set()
+            recipe_dependencies = []
+            try:
+                # Get recipe to add and who's ultimately adding it:
+                recipe = Recipe.get_recipe(name, ctx)
+                recipe_conflicts = {c.lower() for c in recipe.conflicts}
+                recipe_dependencies = get_dependency_tuple_list_for_recipe(
+                    recipe, blacklist=blacklist
+                )
+            except ValueError:
+                pass
+            adder_first_recipe_name = adding_recipe or name
+
+            # Collect the conflicts:
+            triggered_conflicts = []
+            for dep_tuple_list in deps:
+                # See if the new deps conflict with things added before:
+                if len(set(dep_tuple_list).intersection(
+                       recipe_conflicts)) == len(dep_tuple_list):
+                    triggered_conflicts.append(dep_tuple_list)
+                    continue
+
+                # See if what was added before conflicts with the new deps:
+                if len(dep_tuple_list) > 1:
+                    # Not an obvious commitment to a specific recipe/dep
+                    # to be added, so we won't check.
+                    # (remember this function only catches obvious issues)
+                    continue
+                try:
+                    dep_recipe = Recipe.get_recipe(dep_tuple_list[0], ctx)
+                except ValueError:
+                    continue
+                conflicts = [c.lower() for c in dep_recipe.conflicts]
+                if name in conflicts:
+                    triggered_conflicts.append(dep_tuple_list)
+
+            # Throw error on conflict:
+            if triggered_conflicts:
+                # Get first conflict and see who added that one:
+                adder_second_recipe_name = "'||'".join(triggered_conflicts[0])
+                second_recipe_original_adder = deps_were_added_by.get(
+                    adder_second_recipe_name, None
+                )
+                if second_recipe_original_adder:
+                    adder_second_recipe_name = second_recipe_original_adder
```

Comment:
not covered by the tests

Reply:
the function? or what exactly are you referring to?

that error is covered by `def test_invalid_obvious_conflict_checker(names, bootstrap)`

Reply:
I'm saying this specific line `201` is never reached by any of the test. Put a breakpoint and see it yourself

Reply:
ah, fascinating! okay let me think up a test case where it will be hit, I'll add one

Reply:
ahah now we're getting to the point :smile:
So if you can't think of a test, how do you expect another maintainer to understand this code path?
And if there're no possible code path, then it's dead code and you can delete it.
I'm not pretending that I know if there is a path or not. Just pointing that sometimes things are not as obvious as we thought in the first place, that's why we have tests

Reply:
no the problem is just that the available recipes are really way too simple for this. you need to build up some indirection, it's not that the setup per se is complicated, but the current tests quite frankly suck a little in that regard... (no code anywhere to set up some more complex recipe dependencies than the ones the actual recipes use, which just don't cut it)

> then it's dead code

it's not dead code

Reply:
I actually added a test now, `test_indirectconflict_obvious_conflict_checker`. and it had a small bug :joy: only affected the error message, but still!

resolving since it's covered & fixed now